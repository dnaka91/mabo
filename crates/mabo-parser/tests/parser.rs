use std::{
    fmt::{self, Display},
    fs,
    path::{Path, PathBuf},
};

use insta::{assert_snapshot, glob, with_settings};
use miette::{Diagnostic, MietteHandler, MietteHandlerOpts, ReportHandler};
use mabo_parser::Schema;

fn strip_path(path: &Path) -> PathBuf {
    path.strip_prefix(concat!(env!("CARGO_MANIFEST_DIR"), "/tests/inputs"))
        .unwrap()
        .to_owned()
}

#[test]
fn parse_schema() {
    glob!("inputs/*.mabo", |path| {
        let input = fs::read_to_string(path).unwrap();
        let value = Schema::parse(input.as_str(), Some(&strip_path(path))).unwrap();

        with_settings!({
            description => input.trim(),
            omit_expression => true,
        }, {
            assert_snapshot!("parse", format!("{value:#?}"));
            assert_snapshot!("print", value.to_string());
        });
    });
}

#[test]
fn parse_invalid_schema() {
    struct Wrapper<'a>(&'a MietteHandler, &'a dyn Diagnostic);

    impl Display for Wrapper<'_> {
        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
            self.0.debug(self.1, f)
        }
    }

    let handler = MietteHandlerOpts::new()
        .color(false)
        .terminal_links(false)
        .width(120)
        .force_graphical(true)
        .build();

    glob!("inputs/invalid/*.mabo", |path| {
        let input = fs::read_to_string(path).unwrap();
        let value = Schema::parse(input.as_str(), Some(&strip_path(path))).unwrap_err();

        with_settings!({
            description => input.trim(),
            omit_expression => true,
        }, {
            assert_snapshot!("error", Wrapper(&handler, &value).to_string());
        });
    });
}
