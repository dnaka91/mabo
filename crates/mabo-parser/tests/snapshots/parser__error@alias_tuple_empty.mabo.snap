---
source: crates/mabo-parser/tests/parser.rs
description: type Sample = ();
input_file: crates/mabo-parser/tests/inputs/invalid/alias_tuple_empty.mabo
---
mabo::parse::alias_def (https://docs.rs/mabo-parser/0.1.0/mabo_parser/error/struct.ParseAliasError.html)

  × Failed to parse type alias declaration
  ├─▶ Failed to parse type definition
  ├─▶ Failed to parse type definition
  ╰─▶ Parser error at offset 15
   ╭─[invalid/alias_tuple_empty.mabo:1:16]
 1 │ type Sample = ();
   ·                ▲
   ·                ╰── In this declaration
   ╰────
  help: Expected type alias declaration in the form `❬B❭type <Alias> = <Type>;❬B❭`

Error: mabo::parse::type_def (https://docs.rs/mabo-parser/0.1.0/mabo_parser/error/struct.ParseTypeError.html)

  × Failed to parse type definition
  ├─▶ Failed to parse type definition
  ╰─▶ Parser error at offset 15
   ╭─[invalid/alias_tuple_empty.mabo:1:15]
 1 │ type Sample = ();
   ·               ┬
   ·               ╰── In this declaration
   ╰────
  help: Expected type definition in the form `❬B❭<Name>❬B❭`

Error: mabo::parse::type_def (https://docs.rs/mabo-parser/0.1.0/mabo_parser/error/struct.ParseTypeError.html)

  × Failed to parse type definition
  ╰─▶ Parser error at offset 15
   ╭─[invalid/alias_tuple_empty.mabo:1:16]
 1 │ type Sample = ();
   ·                ▲
   ·                ╰── In this declaration
   ╰────
  help: Expected type definition in the form `❬B❭<Name>❬B❭`

Error: 
  × Parser error at offset 15
